pkg/gcv/parallelvalidator.go

```diff
@@ -0,0 +1,111 @@
+package gcv
+
+import (
+	"context"
+	"flag"
+	"runtime"
+
+	"github.com/pkg/errors"
+
+	"github.com/golang/glog"
+
+	"github.com/forseti-security/config-validator/pkg/api/validator"
+	"github.com/forseti-security/config-validator/pkg/multierror"
+)
+
+var flags struct {
+	workerCount int
+}
+
+func init() {
+	flag.IntVar(
+		&flags.workerCount,
+		"workerCount",
+		runtime.NumCPU(),
+		"Number of workers that Validator will spawn to handle validate calls, this defaults to core count on the host")
+}
+
+// ParallelValidator handles making parallel calls to Validator during a Review call.
+type ParallelValidator struct {
+	cv   ConfigValidator
+	work chan func()
+}
+
+type assetResult struct {
+	violations []*validator.Violation
+	err        error
+}
+
+// NewParallelValidator creates a new instance with the given stop channel and validator
+func NewParallelValidator(stopChannel <-chan struct{}, cv ConfigValidator) *ParallelValidator {
+	pv := &ParallelValidator{
+		work: make(chan func(), flags.workerCount*2),
+		cv:   cv,
+	}
+
+	go func() {
+		<-stopChannel
+		glog.Infof("validator shutdown requested via stopChannel close")
+		close(pv.work)
+	}()
+
+	workerCount := flags.workerCount
+	glog.Infof("validator starting %d workers", workerCount)
+	for i := 0; i < workerCount; i++ {
+		go pv.reviewWorker(i)
+	}
+
+	return pv
+}
+
+// reviewWorker is the function that each worker goroutine will use
+func (v *ParallelValidator) reviewWorker(idx int) {
+	glog.V(1).Infof("worker %d starting", idx)
+	for f := range v.work {
+		f()
+	}
+	glog.V(1).Infof("worker %d terminated", idx)
+}
+
+// handleReview is the wrapper function for individual asset reviews.
+func (v *ParallelValidator) handleReview(ctx context.Context, idx int, asset *validator.Asset, resultChan chan<- *assetResult) func() {
+	return func() {
+		resultChan <- func() *assetResult {
+			violations, err := v.cv.ReviewAsset(ctx, asset)
+			if err != nil {
+				return &assetResult{err: errors.Wrapf(err, "index %d", idx)}
+			}
+			return &assetResult{violations: violations}
+		}()
+	}
+}
+
+// Review evaluates each asset in the review request in parallel and returns any
+// violations found.
+func (v *ParallelValidator) Review(ctx context.Context, request *validator.ReviewRequest) (*validator.ReviewResponse, error) {
+	assetCount := len(request.Assets)
+	resultChan := make(chan *assetResult, flags.workerCount*2)
```

Comment:
should the size of resultChan match that of ParallelValidator.work? if so, can we add a comment?

Reply:
done